code/datums/observation/observation.dm

```diff
@@ -121,15 +123,25 @@
 		callbacks = list()
 		listeners[listener] = callbacks
 
-	// If the proc_call is already registered skip
+	if(!event_source.observations)
+		event_source.observations = list()
+	if(!event_source.observations[src])
+		event_source.observations[src] = list()
+	if(!event_source.observations[src][listener])
+		event_source.observations[src][listener] =  list()
+
+	// If the proc_call is already registered skip in Singleton
 	if(proc_call in callbacks)
 		return FALSE
+	if(proc_call in event_source.observations[src][listener][event_source])
+		return FALSE
 
 	// Add the callback, and return true.
 	callbacks += proc_call
+	event_source.observations[src][listener][event_source] += proc_call
 	return TRUE
 
-/decl/observ/proc/unregister(var/event_source, var/datum/listener, var/proc_call)
+/decl/observ/proc/unregister(datum/event_source, datum/listener, proc_call)
 	// Sanity.
 	if (!(event_source && listener && (event_source in event_sources)))
 		return FALSE
```

Comment:
you've not done anything with this, need me to explain how unregister should work?